amethyst_ui/src/pass.rs

```diff
@@ -209,87 +199,373 @@ impl<'a> ParallelIterator for DrawUiApply<'a> {
+                        let end_byte = rendered_string
+                            .grapheme_indices(true)
+                            .nth(end)
+                            .map(|i| i.0)
+                            .unwrap_or(rendered_string.len());
+                        start_byte.map(|start_byte| (editing, (start_byte, end_byte)))
+                    })
+                    .map(|(editing, (start_byte, end_byte))| {
+                        vec![
+                            SectionText {
+                                text: &((rendered_string)[0..start_byte]),
+                                scale: Scale::uniform(ui_text.font_size),
+                                color: ui_text.color,
+                                font_id: FontId(0),
+                            },
+                            SectionText {
+                                text: &((rendered_string)[start_byte..end_byte]),
+                                scale: Scale::uniform(ui_text.font_size),
+                                color: editing.selected_text_color,
+                                font_id: FontId(0),
+                            },
+                            SectionText {
+                                text: &((rendered_string)[end_byte..]),
+                                scale: Scale::uniform(ui_text.font_size),
+                                color: ui_text.color,
+                                font_id: FontId(0),
+                            },
+                        ]
+                    })
+                    .unwrap_or(vec![
+                        SectionText {
+                            text: rendered_string,
+                            scale: Scale::uniform(ui_text.font_size),
+                            color: ui_text.color,
+                            font_id: FontId(0),
+                        },
+                    ]);
+                // TODO: If you're adding multi-line support you need to change this to use
+                // Layout::Wrap.
+                let layout = Layout::SingleLine {
+                    line_breaker: BuiltInLineBreaker::UnicodeLineBreaker,
+                    h_align: HorizontalAlign::Left,
+                    v_align: VerticalAlign::Top,
+                };
+                let section = VariedSection {
+                    screen_position: (ui_transform.x, ui_transform.y),
+                    bounds: (ui_transform.width, ui_transform.height),
+                    z: ui_transform.z,
+                    layout,
+                    text,
+                };
 
-                    if let Some(image) = ui_text
-                        .get(entity)
-                        .and_then(|ref ui_text| ui_text.texture.as_ref())
-                        .and_then(|texture| tex_storage.get(texture))
+                // Render background highlight
+                let brush = &mut self.glyph_brushes
+                    .get_mut(&ui_text.brush_id.unwrap())
+                    .unwrap()
```

Comment:
Are all these unwraps safe?

Reply:
yes they are, because the code beforehand proved the unwrap safe.  The only reason option values are used here is because the structures can't be populated at initialization.